### PR TITLE
Implement kubernetes endpoints

### PR DESCRIPTION
**Pull request by Ntr0**

### CODE REVIEW

File: k8s.go [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+	// Enum: [k8s]
+	PBType string `json:"type,omitempty"`
+
+	// Entities of a cluster
+	Entities KubernetesClusterEntities `json:"entities"`
+}
+
+type KubernetesClusterEntities struct {
+	// NodePools of a cluster
+	NodePools *KubernetesNodePools `json:"nodepools,omitempty"`
```
Comment (piepmatz) [inline k8s.go]: `KubernetesNodePools` has `Items`. According to https://devops.ionos.com/api/cloud/v5/#read-cluster this is not the case for a cluster's nodepool entities.

File: k8s.go [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+	// URL to the collection representation (absolute path)
+	// Read Only: true
+	// Format: uri
+	Href string `json:"href,omitempty"`
+
+	// Unique representation for Kubernetes Cluster as a collection on a resource.
+	// Read Only: true
+	ID string `json:"id,omitempty"`
+
+	// Array of items in that collection
```
Comment (piepmatz) [inline k8s.go]: It's a Go SDK, so it's a slice

File: k8s.go [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+	// Read Only: true
+	Items []*KubernetesCluster `json:"items"`
+
+	// The type of resource within a collection
+	// Read Only: true
+	// Enum: [k8s]
```
Comment (piepmatz) [inline k8s.go]: https://devops.ionos.com/api/cloud/v5/#list-clusters says `collection`

File: k8s.go [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+
+type KubernetesClusters struct {
+	// URL to the collection representation (absolute path)
+	// Read Only: true
+	// Format: uri
+	Href string `json:"href,omitempty"`
+
+	// Unique representation for Kubernetes Cluster as a collection on a resource.
+	// Read Only: true
+	ID string `json:"id,omitempty"`
```
Comment (piepmatz) [inline k8s.go]: Would your `// Enum: [k8s]` comment make sense here?

File: k8s.go [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+
+	// Array of items in that collection
+	// Read Only: true
+	Items []*KubernetesCluster `json:"items"`
```
Comment (piepmatz) [inline k8s.go]: - I haven't understood yet when you use pointers and when you don't. Can you explain?
- Same as for a cluster's nodepool entities: The things in `Items` are not full-fledged clusters.

File: k8s.go [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+	// Read Only: true
+	// Format: uri
+	Href string `json:"href,omitempty"`
+
+	// The resource's unique identifier.
+	// Read Only: true
+	ID string `json:"id,omitempty"`
+
+	// metadata
+	Metadata *Metadata `json:"metadata,omitempty"`
```
Comment (piepmatz) [inline k8s.go]: this thing lacks the `createdByUserId` and `lastModifiedByUserId` fields

File: k8s.go [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+
+	// Number of cores for node
+	// Required: true
+	CoresCount int64 `json:"coresCount,omitempty"`
+
+	// A valid cpu family name
+	// Required: true
+	CPUFamily string `json:"cpuFamily,omitempty"`
+
+	// A valid uuid of the datacenter on which user has access
```
Comment (piepmatz) [inline k8s.go]: Yeah, but mainly where the nodepool is provisioned. The official description "The unique identifier of the data center where the worker nodes of the node pool are provisioned." fits quite well.

File: k8s.go [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+// swagger:model KubernetesNodePoolProperties
+type KubernetesNodePoolProperties struct {
+	// The availability zone in which the server should exist
+	// Required: true
+	// Enum: [AUTO ZONE_1 ZONE_2]
+	AvailabilityZone string `json:"availabilityZone,omitempty"`
+
+	// Number of cores for node
+	// Required: true
+	CoresCount int64 `json:"coresCount,omitempty"`
```
Comment (piepmatz) [inline k8s.go]: `int64`? Wouldn't `int32` or `uint32` have been more than enough?

File: k8s.go [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+	// The type of object
+	// Read Only: true
+	// Enum: [nodepool]
+	PBType string `json:"type,omitempty"`
+}
+
+// KubernetesNodePoolProperties kubernetes node pool properties
+// swagger:model KubernetesNodePoolProperties
+type KubernetesNodePoolProperties struct {
+	// The availability zone in which the server should exist
```
Comment (piepmatz) [inline k8s.go]: server**s**. or rather "nodes" in this context?

File: k8s.go [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+}
+
+type KubernetesClusterEntities struct {
+	// NodePools of a cluster
+	NodePools *KubernetesNodePools `json:"nodepools,omitempty"`
+}
+
+// KubernetesClusterProperties kubernetes cluster properties
+type KubernetesClusterProperties struct {
+	// A Kubernetes Cluster Name. Valid Kubernetes Cluster name must be 63 characters or less and must be empty
```
Comment (piepmatz) [inline k8s.go]: As the field is required I am surprised that "empty" should be valid. Have you tried?

File: k8s.go [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+	// Read Only: true
+	// Format: uri
+	Href string `json:"href,omitempty"`
+
+	// The resource's unique identifier.
+	// Read Only: true
+	ID string `json:"id,omitempty"`
+
+	// metadata
+	Metadata *NoStateMetaData `json:"metadata,omitempty"`
```
[the remark this answers is not in the document]
Comment (Ntr0) [reply k8s.go]: Well, I was lazy and generated the models using go swagger. I'll look over them again

File: shared_types.go [hunk 12]
```diff
@@ -0,0 +1,39 @@
+package profitbricks
+
+import (
+	"github.com/go-openapi/strfmt"
+)
+
+// NoStateMetaData no state meta data
+type NoStateMetaData struct {
```
[the remark this answers is not in the document]
Comment (Ntr0) [reply shared_types.go]: please have a look here: https://api.ionos.com/cloudapi/v5/swagger.json

[the remark this answers is not in the document]
Comment (Ntr0) [reply shared_types.go]: There seems to be a mismatch between swagger, api doc and real life. I'll follow api doc then.

File: k8s.go (hunk 1 above)
Comment (Ntr0) [reply k8s.go]: Probably the example in the doc was not with increased depth. Please check the examples from testdata, they are based on real life responses

File: k8s.go (hunk 5 above)
Comment (Ntr0) [reply k8s.go]: - swagger generate decided to do it like this. But I'll change it to match the other models.
- They are, if the depth is high enough

File: k8s.go (hunk 10 above)
Comment (Ntr0) [reply k8s.go]: Empty is not valid, changed comment

File: k8s.go (hunk 4 above)
Comment (Ntr0) [reply k8s.go]: I do not understand your comment.

File: k8s.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (piepmatz) [reply k8s.go]: I'd still say it's not writable and therefore "Read Only" :)

File: k8s.go [hunk 13]
```diff
@@ -0,0 +1,247 @@
...
+	Properties KubernetesNodePoolProperties `json:"properties"`
+
+	// The type of object
+	// Read Only: true
+	// Enum: [nodepool]
+	PBType string `json:"type,omitempty"`
+}
+
+// KubernetesNodePoolProperties kubernetes node pool properties
+// swagger:model KubernetesNodePoolProperties
```
Comment (piepmatz) [inline k8s.go]: Remove those swagger comments (here and for KubernetesNodePools)

File: shared_types.go [hunk 14]
```diff
@@ -0,0 +1,39 @@
+package profitbricks
+
+import (
+	"github.com/go-openapi/strfmt"
+)
+
+// NoStateMetaData no state meta data
```
[the remark this answers is not in the document]
Comment (piepmatz) [reply shared_types.go]: Honestly, "// KubernetesNodePoolProperties kubernetes node pool properties" is just not useful at all. Better just remove it if you don't think an exported something needs an explanation.

File: k8s.go (hunk 8 above)
Comment (piepmatz) [reply k8s.go]: There are still some int64 occurrences around (NodeCount, RAMSize, StorageSize). For all of them uint32 seems fine.

File: k8s.go [hunk 15]
```diff
@@ -0,0 +1,247 @@
...
+	// Format: uri
+	Href string `json:"href,omitempty"`
+
+	// The resource's unique identifier.
+	// Read Only: true
+	ID string `json:"id,omitempty"`
+
+	// properties
+	// Required: true
+	Properties *KubernetesConfigProperties `json:"properties"`
```
[the remark this answers is not in the document]
Comment (Ntr0) [reply k8s.go]: you are right, the kubeconfig will be delivered with depth=0, so there is no need to use a pointer. Thats different for the other properties, so I changed to pointers where appropriate.

File: k8s.go [hunk 16]
```diff
@@ -0,0 +1,247 @@
...
+// GetKubernetesCluster gets cluster with given id
+func (c *Client) GetKubernetesCluster(clusterId string) (*KubernetesCluster, error) {
+	rsp := &KubernetesCluster{}
+	return rsp, c.GetOK(kubernetesClusterPath(clusterId), rsp)
+}
+
+// CreateKubernetesCluster creates a cluster
+func (c *Client) CreateKubernetesCluster(cluster KubernetesCluster) (*KubernetesCluster, error) {
+	rsp := &KubernetesCluster{}
+	return rsp, c.PostAcc(kubernetesClustersPath(), cluster, rsp)
```
[the remark this answers is not in the document]
Comment (Ntr0) [reply k8s.go]: tried it, docs are wrong

File: k8s.go [hunk 17]
```diff
@@ -130,7 +129,7 @@ type KubernetesNodePoolProperties struct {
 
 	// The unique identifier of the data center where the worker nodes of the node pool will be provisioned.
 	// Required: true
-	DatacenterID string `json:"datacenterId,omitempty"`
```
Comment (piepmatz) [inline k8s.go]: this was right, now it's wrong

File: k8s.go [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+
+type KubernetesClusterEntities struct {
+	// NodePools of a cluster
+	NodePools *KubernetesNodePools `json:"nodepools,omitempty"`
+}
+
+// KubernetesClusterProperties kubernetes cluster properties
+type KubernetesClusterProperties struct {
+	// A Kubernetes Cluster Name. Valid Kubernetes Cluster name must be 63 characters or less and must not be empty
+	// and begin and end with an alphanumeric character ([a-z0-9A-Z]) with dashes (-), dots (.),
```
Comment (piepmatz) [inline k8s.go]: I looked up the underscore issue again. Underscores are fine, the docs are correct.

You confused the restrictions applying to label/annotation **keys' prefix parts**.  The keys' name parts can include underscores. The values, too.
The restriction applying to the values are identical to the restrictions for cluster and nodepool names, so underscores are fine.
(Please also think of informing Ionos again that your previous information about underscores being invalid was not correct.)

Comment (piepmatz) [reply k8s.go]: There's something I need to add: The cluster and nodepool names were intended to be DNS labels but the regex in place allows too much, making the names actually K8s qualified names. So underscores are allowed here, even though DNS wouldn't allow them.

**Pull request merged**